unifyfs utility to wait until all servers become ready.

sandrain opened the pull request:
unifyfs utility to wait until all servers become ready.
- unifyfsd publishes a server pid file (unifyfsd.pids) under the shared
state state directory, indicating that all servers are ready
- unifyfs utility blocks until the pid file appears

Also, this includes an additional commit that fixes #505.

TEST_CHECKPATH_SKIP_FILES=common/src/unifyfs_configurator.h

<!--- Provide a general summary of your changes in the Title above -->

### Description
This addresses #499.



### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Testing (addition of new tests or update to current tests)
- [ ] Documentation (a change to man pages or other documentation)

MichaelBrim commented on server/src/unifyfs_server_pid.c:
```diff
@@ -0,0 +1,207 @@
+            LOGDBG("all servers are ready to accept client connection");
+        }
+
+out_free:
```
instead of having two labels at the end of this function, just use one like this:
```
out:
    if (NULL != server_pids) {
        free(server_pids);
    }
    return  ret;
```

MichaelBrim commented on server/src/unifyfs_server_pid.c:
```diff
@@ -0,0 +1,207 @@
+}
+DEFINE_MARGO_RPC_HANDLER(server_pid_handle_rpc);
+
+static int set_timeout(void)
```
suggest making this 'static inline'

sandrain commented:
Not sure why spack fails during CI build:

```
before_install.1
4610.00s$ test -n $CC && unset CC
before_install.2
46212.42s$ (cd $HOME/spack; git describe) || git clone https://github.com/spack/spack $HOME/spack
before_install.3
4710.00s$ test -f $HOME/spack/etc/spack/packages.yaml || cat > $HOME/spack/etc/spack/packages.yaml << '  EOF'
4970.53s$ $HOME/spack/bin/spack bootstrap
498==> Error: bootstrap is not a recognized Spack command or extension command; check with `spack commands`.
499The command "$HOME/spack/bin/spack bootstrap" failed and exited with 1 during .
```

MichaelBrim commented on server/src/unifyfs_server_pid.c:
```diff
@@ -0,0 +1,207 @@
+    int ret = 0;
+    char filename[PATH_MAX] = { 0, };
+    FILE* fp = NULL;
+
```
even though this should only be called when `server_pids` has been setup correctly, it's still better to check that it's not NULL before accessing it

MichaelBrim commented on util/unifyfs/src/unifyfs-rm.c:
```diff
@@ -807,11 +869,25 @@ int unifyfs_start_servers(unifyfs_resource_t* resource,
-        return resource_managers[resource->rm].launch(resource, args);
+    pid = fork();
+    if (pid < 0) {
+        fprintf(stderr, "failed to launch the server process (%s)\n",
```
error message isn't quite accurate. how about "failed to create the server launch process"?

MichaelBrim commented on server/src/unifyfs_server_pid.c:
```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2017, Lawrence Livermore National Security, LLC.
```
suggest updating copyright year to 2020 for new files

MichaelBrim commented on server/src/unifyfs_global.h:
```diff
@@ -197,5 +198,13 @@ unifyfs_rc disconnect_app_client(app_client* clnt);
 
 unifyfs_rc cleanup_app_client(app_client* clnt);
 
+/**
```
Since `unifyfs_publish_server_pids()` is only used in `server/src/unifyfs_server.c`, I suggest you move this declaration to that file.

sandrain replied:
I initially did that but guess checkpatch script was not happy. I will check it again.

sandrain commented:
@MichaelBrim Thanks for comments. I will make changes accordingly.

CamStan commented on common/src/unifyfs_configurator.h:
```diff
@@ -90,6 +90,7 @@
     UNIFYFS_CFG_CLI(runstate, dir, STRING, RUNDIR, "runstate file directory", configurator_directory_check, 'R', "specify full path to directory to contain server runstate file") \
     UNIFYFS_CFG_CLI(server, hostfile, STRING, NULLSTRING, "server hostfile name", NULL, 'H', "specify full path to server hostfile") \
     UNIFYFS_CFG_CLI(sharedfs, dir, STRING, NULLSTRING, "shared file system directory", configurator_directory_check, 'S', "specify full path to directory to contain server shared files") \
+    UNIFYFS_CFG_CLI(server, init_timeout, INT, UNIFYFS_DEFAULT_INIT_TIMEOUT, "timeout of waiting for server initialization", NULL, 't', "timeout in seconds to wait for servers to be ready for clients") \
```
Mind adding this to our docs ([here](https://unifyfs.readthedocs.io/en/latest/configuration.html#id7)) with a note on what the default is?

E.g.: timeout in seconds to wait for servers to be ready for clients (default: 120)

sandrain replied:
@CamStan I will add it

MichaelBrim commented on server/src/unifyfs_server_pid.c:
```diff
@@ -0,0 +1,213 @@
+    } else {
+        ret = set_timeout();
+        if (ret) {
+            goto out;
```
no need for goto here, just `return ret`

MichaelBrim commented on server/src/unifyfs_server_pid.c:
```diff
@@ -0,0 +1,213 @@
+        server_pids = calloc(glb_pmi_size, sizeof(*server_pids));
+        if (!server_pids) {
+            LOGERR("failed to allocate memory (%s)", strerror(errno));
+            goto out;
```
no need for goto here, just `return ENOMEM`

MichaelBrim commented on server/src/unifyfs_server_pid.c:
```diff
@@ -0,0 +1,213 @@
+
+    fclose(fp);
+
+out:
```
you should only use labels when there is common code that must be run at the end of the function in many cases. this function has no such code. replace all your gotos with explicit returns.

sandrain commented:
@MichaelBrim Thanks. I've removed unnecessary goto statements. I've also added code to unlink existing pid file (somehow if any) before launch the servers.

I am wondering if there exist any use-case that a user starts unifyfsd twice in a single job? For instance:

```
unifyfs start 
jsrun/srun /some/app
unifyfs terminate

unifyfs start
jsrun /some/other/independent/app
unifyfs terminate
```

I am just thinking if we need to have cleanup routines (shm, etc. #496) when we terminate the server.

MichaelBrim commented:
@sandrain, regarding cleanup. When the servers shutdown cleanly they will remove all the state correctly. It's the failed server runs (e.g., segfaults) that are problematic. For your example, I assume the first instance ran successfully, so there should be no issue starting up the second instance.

MichaelBrim commented on server/src/unifyfs_server_pid.c:
```diff
@@ -0,0 +1,210 @@
+                        &handle);
+    if (hret != HG_SUCCESS) {
+        LOGERR("failed to create rpc handle (ret=%d)", hret);
+        return EIO;
```
These `EIO` return codes should be `UNIFYFS_ERROR_MARGO`.

MichaelBrim commented on server/src/unifyfs_server_pid.c:
```diff
@@ -0,0 +1,210 @@
+    }
+out:
+    if (server_pids) {
+        free(server_pids);
```
After free(), you should set `server_pids` back to NULL.

sandrain commented:
@MichaelBrim I have included an additional commit to fix #505.
